### ISSUE TITLE
Inverse VQT, griffinlim-vqt

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#1161]
**Is your feature request related to a problem? Please describe.**

#1018 implements variable-Q transforms.  This issue exists to factor out the related methods for inverse-VQT and VQT phase recovery.  These are not critical to implementing VQT, and can be punted to a later release.

**Comment by lostanlen** [comment]
I agree. I don't mean to sound pessimistic, but there's a chance that our VQT implementation is slightly buggy and has weird corner cases. Let's wait till those bugs get reported and fixed before worrying about feature inversion.

**Comment by bmcfee** [comment]
This is probably worth investigating again since merging #1405.

**Comment by bmcfee** [comment]
Ok, I think i've tracked down the source of the scaling error that I've been running into with vqt inversion, and it's pretty subtle.

I noticed when implemented #1405 that the length normalization in the forward VQT here:
https://github.com/librosa/librosa/blob/8bdfc032d1b362a4c29becbee8dc89f082e5fb34/librosa/core/constantq.py#L992-L1006

needs to recompute the filter lengths at the end because the sampling rate may have changed by early downsampling:
https://github.com/librosa/librosa/blob/8bdfc032d1b362a4c29becbee8dc89f082e5fb34/librosa/core/constantq.py#L917-L919

In the CQT case (gamma=0), this doesn't matter because the resampling factor cancels out, so lengths/sr is the same.  However, when `gamma != 0`, it becomes a problem.  Making this work correctly will require inferring the effective sampling rate of the forward vqt inside the inverse vqt.  This seems a little dicey to me, but it's approximately doable.

**Comment by bmcfee** [comment]
Compensating for early downsampling did not make much of a difference here.  I have something that more or less works, but it's still not giving a better inverse than using `gamma=0` even when a nonzero gamma was used in the forward transform.

**Comment by bmcfee** [comment]
The plot thickens.

It looks like our inverse cqt consistently over-amplifies by 0.5 dB.

```python
>>> sr = 44100
>>> y, sr = librosa.load(librosa.ex('nutcracker'), sr=sr, duration=10, offset=5)

# VQT with gamma=0 ==> cqt
>>> V = librosa.vqt(y=y, sr=sr, gamma=0, hop_length=128, sparsity=0, bins_per_octave=60, n_bins=5*12*8, scale=False, res_type='soxr_hq')

# Inverse CQT
>>> yi = librosa.icqt(C=V, sr=sr, length=len(y), hop_length=128, bins_per_octave=60, sparsity=0, scale=False, res_type='soxr_hq')

# Evaluate residual error
>>> np.sqrt(np.mean((y-yi)**2))
0.0029379776151615493

# Residual error with a 0.5 dB compensation
>>> np.sqrt(np.mean((y-yi* librosa.db_to_amplitude(-0.5))**2))
0.0019134708217162944

# Compute another forward pass
>>> V1 = librosa.vqt(y=yi, sr=sr, gamma=0, hop_length=128, sparsity=0, bins_per_octave=60, n_bins=V.shape[0], scale=False, res_type='soxr_hq')

# Compare amplitude
>>> db_rel1 = librosa.amplitude_to_db(np.abs(V1)/np.abs(V), ref=1, top_db=None)

# Aggregate over time
>>> plt.plot(np.median(db_rel1, axis=1))
```
Results in the following plot:
![image](https://user-images.githubusercontent.com/1190540/143690436-33331e90-1afb-4825-984a-582aa75676f5.png)

This behavior is consistent across configuration parameters (frequency range, resolution, etc).  It's also the same with or without `scale=True` - I'd disabled scaling in the above example to simplify the code path.  It's probably due to some kind of off-by-one error in the per-octave basis scaling.  I'll investigate more.

**Comment by bmcfee** [comment]
Following up on the above, the gain induced by cqt round-trip appears to be window-dependent.  Hann consistently produces 0.5dB, other windows give other values. Probably this is down to using L1-normalized basis functions and then treating the conjugate-transpose as if it was the proper basis inverse.

**Comment by bmcfee** [comment]
The following snippet handles the CQT (gamma=0) case.  The window-dependent gain turned out to be because we weren't accounting for the basis functions' power in the frequency domain.  Fixing this gets us consistently at ~0.001 dB gain, though some windows behave better than others.

I've also simplified the code a bit by making better use of einstein summations.

Two things still bother me:

1. We now get over-amplification of low frequencies when `gamma>0`, so I don't think the inverse VQT issue is resolved yet.
2. There's some scooping up in the top octave that I don't fully grok yet.  This might just be down to filter scaling and hop length.  

That said, the patch below does consistently improve cqt accuracy, so I'll prep a PR as a bugfix to `icqt` for now.

I'm also wondering if it might be worth streamlining the vqt implementation and removing the special case for the top octave.  It makes the code more complicated than I think it needs to be, and the cost benefits seem negligible.

<details>

<summary> Inverse CQT < /summary>

```python
def ivqt(
    C,
    sr=22050,
    hop_length=512,
    fmin=None,
    bins_per_octave=12,
    tuning=0.0,
    filter_scale=1,
    norm=1,
    sparsity=0.01,
    window="hann",
    scale=True,
    length=None,
    res_type="fft",
    gamma=0,
    dtype=None,
):
    if fmin is None:
        fmin = note_to_hz("C1")

    # Apply tuning correction
    fmin = fmin * 2.0 ** (tuning / bins_per_octave)

    # Get the top octave of frequencies
    n_bins = C.shape[-2]

    # How many octaves do we have?
    n_octaves = int(np.ceil(float(n_bins) / bins_per_octave))
    
    # Get the frequencies and filter lengths
    freqs = cqt_frequencies(
        fmin=fmin, n_bins=n_bins, bins_per_octave=bins_per_octave
    )
    alpha = __bpo_to_alpha(bins_per_octave)
    
    lengths, f_cutoff = filters.wavelet_lengths(
        freqs, sr=sr, window=window, filter_scale=filter_scale, alpha=alpha, gamma=gamma,
    )
    
    # Trim the CQT to only what's necessary for reconstruction
    if length is not None:
        n_frames = int(np.ceil((length + max(lengths)) / hop_length))
        C = C[..., :n_frames]

    if scale:
        C_scale = np.sqrt(lengths)
    else:
        C_scale = np.ones_like(lengths)
    
    y = None

    # Assume the top octave is at the full rate
    srs = [sr]
    hops = [hop_length]

    for i in range(n_octaves - 1):
        if hops[0] % 2 == 0:
            # We can downsample:
            srs.insert(0, srs[0] * 0.5)
            hops.insert(0, hops[0] // 2)
        else:
            # We're out of downsamplings, carry forward
            srs.insert(0, srs[0])
            hops.insert(0, hops[0])

    for i, (my_sr, my_hop) in enumerate(zip(srs, hops)):

        # How many filters are in this octave?
        n_filters = min(bins_per_octave, n_bins - bins_per_octave * i)

        # Make a slice for the current octave
        sl = slice(bins_per_octave * i, bins_per_octave * i + n_filters)

        fft_basis, n_fft, _ = __vqt_filter_fft(
            my_sr,
            freqs[sl],
            filter_scale,
            norm,
            sparsity,
            window=window,
            dtype=dtype,
            gamma=gamma,
            alpha=alpha,
        )

        # Transpose the basis
        inv_basis = fft_basis.H.todense()
        
        # Compute each filter's frequency-domain power
        freq_power = 1/np.sum(np.abs(np.asarray(inv_basis))**2, axis=0)
        
        # Compensate for length normalization in the forward transform
        freq_power *= n_fft / lengths[sl] 

        # Inverse-project the basis for each octave
        D_oct = np.einsum("fc,c,c,...ct->...ft", inv_basis, C_scale[sl], freq_power, C[..., sl, :], optimize=True)

        y_oct = istft(D_oct, window="ones", hop_length=my_hop, dtype=dtype)

        y_oct = audio.resample(
            y_oct, 1, sr // my_sr, res_type=res_type, scale=False, fix=False
        )

        if y is None:
            y = y_oct
        else:
            y[..., : y_oct.shape[-1]] += y_oct
            
    if length:
        y = util.fix_length(y, length)

    return y
```

</details>

**Comment by bmcfee** [comment]
Following up on this after spending a bit of time doing cleanup on vqt and icqt in #1407.  The TLDR is that I think we should punt VQT inversion.

Our inverse CQT implementation works by transposing and normalizing the filter bank for each octave, projecting from the CQT basis up to the DFT basis, and then using an istft to overlap-add the individual frames and resynthesize the signal.

For VQT, transposing the filters doesn't work exactly (due to variable width), which results in an over-emphasis of low frequencies.  In theory, a pseudo-inverse (instead of a transpose) could work here.  However, initial experiments demonstrated that the svd-based pseudo-inverse provided by numpy does not produce (inverse) basis functions which are suitably compact in the frequency domain, and the result is pretty terrible.

Nobody seems to be clamoring for this feature anyway, so I'm happy to punt it.  It also turns out that `icqt(vqt(gamma!=0))` (that is, using gamma=0 for synthesis and gamma!=0 for analysis) works pretty well.  Not perfectly, there is definitely some attenuation in the bass frequencies, but it's not garbage either and is perfectly usable for debugging purposes at least.

We may revisit this in the future, but for now I'm not seeing an easy solution.

**Comment by lostanlen** [comment]
I agree. Should we close the issue or leave it open?

**Comment by bmcfee** [comment]
Let's leave it open in case a solution presents itself.
